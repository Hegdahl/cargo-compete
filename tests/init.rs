pub mod common;

use ignore::overrides::Override;
use insta::{assert_json_snapshot, assert_snapshot};
use std::str;

#[test]
fn atcoder_no_crate() -> anyhow::Result<()> {
    let (output, tree) = run("atcoder", ".", "1\n")?;
    assert_snapshot!("atcoder_no_crate_output", output);
    assert_json_snapshot!("atcoder_no_crate_file_tree", tree);
    Ok(())
}

#[test]
fn atcoder_use_crate() -> anyhow::Result<()> {
    let (output, tree) = run("atcoder", ".", "2\n")?;
    assert_snapshot!("atcoder_use_crate_output", output);
    assert_json_snapshot!("atcoder_use_crate_file_tree", tree);
    Ok(())
}

#[test]
fn atcoder_use_crate_via_bianry() -> anyhow::Result<()> {
    let (output, tree) = run("atcoder", ".", "3\n")?;
    assert_snapshot!("atcoder_use_crate_via_bianry_output", output);
    assert_json_snapshot!("atcoder_use_crate_via_bianry_file_tree", tree);
    Ok(())
}

#[test]
fn codeforces() -> anyhow::Result<()> {
    let (output, tree) = run("codeforces", ".", "")?;
    assert_snapshot!("codeforces_output", output);
    assert_json_snapshot!("codeforces_file_tree", tree);
    Ok(())
}

#[test]
fn codeforces_with_path() -> anyhow::Result<()> {
    let (output, tree) = run("codeforces", "codeforces", "")?;
    assert_snapshot!("codeforces_with_path_output", output);
    assert_json_snapshot!("codeforces_with_path_file_tree", tree);
    Ok(())
}

fn run(
    platform: &str,
    path: &str,
    input: &'static str,
) -> anyhow::Result<(String, serde_json::Value)> {
    common::run(
        |_| Ok(()),
        input.as_bytes(),
        &["", "compete", "i", platform, path],
        |cwd, output| {
            output
                .replace(&*cwd.to_string_lossy(), "{{ cwd }}")
                .replace(std::path::MAIN_SEPARATOR, "{{ main_path_separator }}")
        },
        |_| Ok(Override::empty()),
    )
}
