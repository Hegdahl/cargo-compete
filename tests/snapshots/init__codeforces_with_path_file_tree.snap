---
source: tests/init.rs
expression: tree
---
{
  "codeforces": {
    ".cargo": {
      "config.toml": "[build]\ntarget-dir = \"target\"\n"
    },
    "compete.toml": "# Path to the test file (Liquid template)\n#\n# Variables:\n#\n# - `manifest_dir`: Package directory\n# - `contest`:      Contest ID (e.g. \"abc100\")\n# - `bin_name`:     Name of a `bin` target (e.g. \"abc100-a\")\n# - `bin_alias`:    \"Alias\" for a `bin` target defined in `pacakge.metadata.cargo-compete` (e.g. \"a\")\n# - `problem`:      Alias for `bin_alias` (deprecated)\n#\n# Additional filters:\n#\n# - `kebabcase`: Convert to kebab case (by using the `heck` crate)\ntest-suite = \"{{ manifest_dir }}/testcases/{{ bin_alias }}.yml\"\n\n# Open files with the command (`jq` command that outputs `string[] | string[][]`)\n#\n# VSCode:\n#open = '[[\"code\", \"-a\", .manifest_dir], [\"code\"] + (.paths | map([.src, .test_suite]) | flatten)]'\n# Emacs:\n#open = '[\"emacsclient\", \"-n\"] + (.paths | map([.src, .test_suite]) | flatten)'\n\n[new]\nkind = \"cargo-compete\"\n# Platform\n#\n# - atcoder\n# - codeforces\n# - yukicoder\nplatform = \"codeforces\"\n# Path (Liquid template)\n#\n# Variables:\n#\n# - `contest`:      Contest ID. **May be nil**\n# - `package_name`: Package name\npath = \"./{{ contest }}\"\n\n#[new]\n#kind = \"oj-api\"\n#url = \"https://atcoder.jp/contests/{{ id }}\"\n#path = \"./{{ contest }}\"\n\n[new.template]\n#lockfile = \"/path/to/cargo-lock-template.toml\"\n# `profile` for `Cargo.toml`.\n#\n# By setting this, you can run tests with `opt-level=3` while enabling `debug-assertions` and `overflow-checks`.\n#profile = '''\n#[dev]\n#opt-level = 3\n#'''\n\n[new.template.dependencies]\nkind = \"inline\"\ncontent = '''\n#proconio = { version = \"0.3.6\", features = [\"derive\"] }\n'''\n\n[new.template.src]\nkind = \"inline\"\ncontent = '''\nfn main() {\n    todo!();\n}\n'''\n\n[test]\n# Profile for `cargo build`. (\"dev\" | \"release\")\n#\n# Defaults to `\"dev\"`.\n#profile = \"dev\"\n\n#[submit.transpile]\n#kind = \"command\"\n#args = [\"cargo\", \"equip\", \"--resolve-cfgs\", \"--remove\", \"docs\", \"--minify\", \"libs\", \"--rustfmt\", \"--check\", \"--bin\", \"{{ bin_name }}\"]\n##language_id = \"\"",
    "rust-toolchain": "1.42.0"
  }
}
